src/main/scala/rps/Game.scala

```diff
@@ -2,69 +2,64 @@ package rps
+import rps.models.Move.Scissors
 
 object Game {
 
-  // TODO this should be an enum (or Scala equivalent of TS Enum)
-  object Weapon {
-    val Rock: String = "0" // 🪨
-    val Paper: String = "1" // 📄
-    val Scissors: String = "2" // ✂️
-  }
-
-  // TODO this should be an enum (or Scala equivalent of TS Enum)
-  object Winner {
-    val Draw: String = "0" // It's a draw
-    val User: String = "1" // User wins
-    val Cpu: String = "2" // Cpu wins
-  }
-
   def play(): Unit = {
-    println(s"""Choose your weapon:
-    ${Weapon.Rock} - ${matchWeapon(Weapon.Rock)}
-    ${Weapon.Paper} - ${matchWeapon(Weapon.Paper)}
-    ${Weapon.Scissors} - ${matchWeapon(Weapon.Scissors)}
-    """)
-    val rawUserInput = readLine()
-    val userWeapon = rawUserInput
-    val cpuWeapon = generateComputerMove()
+    val menu = Move.moves
+      .map(m => s"${Move.encode(m)} - ${Move.print(m)}")
+      .mkString("\n")
+    println("Choose your weapon:")
+    println(menu)
 
-    val result = checkWinner(userWeapon, cpuWeapon)
+    val rawUserInput = readLine()
+    val userMove = Move.decode(rawUserInput)
+    val cpuMove = generateComputerMove()
 
-    println(s"You chose: ${matchWeapon(userWeapon)} | your opponent chose: ${matchWeapon(cpuWeapon)}")
-    println(announceWinner(result))
-  }
+    val printedUserMove =
+      userMove.map(Move.print).getOrElse("💩 An invalid weapon")
+    val printedCpuMove = Move.print(cpuMove)
+    println(
+      s"You chose: ${printedUserMove} | your opponent chose: ${printedCpuMove}"
+    )
 
-  def matchWeapon(weapon: String): String = {
-    weapon match {
-      case Weapon.Rock     => "🪨 Rock"
-      case Weapon.Paper    => "📄 Paper"
-      case Weapon.Scissors => "✂️ Scissors"
-      case _               => "💩 An invalid weapon"
-    }
+    val result = checkWinner(userMove, cpuMove)
+    println(announceResult(result))
   }
 
-  def announceWinner(result: String): String = {
+  def announceResult(result: GameResult): String = {
     result match {
-      case Winner.Draw => "It's a draw ✍️"
-      case Winner.User => "You win 🎉"
-      case _           => "You lose 🤷"
+      case GameResult.Draw     => "It's a draw ✍️"
+      case GameResult.UserWins => "You win 🎉"
+      case GameResult.CpuWins  => "You lose 🤷"
+      case GameResult.DumbUser =>
+        "You lose 🤷 (next time choose a valid weapon)"
     }
   }
 
-  def checkWinner(userMove: String, computerMove: String): String = {
+  def checkWinner(
+      userMove: Option[Move],
+      computerMove: Move
+  ): GameResult = {
     (userMove, computerMove) match {
-      case (x, y) if x == y              => Winner.Draw
-      case (Weapon.Rock, Weapon.Scissors)  => Winner.User
-      case (Weapon.Paper, Weapon.Rock)     => Winner.User
-      case (Weapon.Scissors, Weapon.Paper) => Winner.User
-      case _                               => Winner.Cpu
-      // default case also handles an invalid weapon (e.g. user inputs "4")
+      case (Some(x), y) =>
```

Comment:
suggerirei di fare il primo match solo su `userMove`, perché tanto stai ignorando `computerMove`

poi quello dentro lo fai su `(x, computerMove)`

Reply:
visto che parlavamo dell'API di Option ieri, questo è un caso in cui (una volta riscritto come suggerisco), si potrebbe anche evitare il match esterno e fare
```scala
userMove.map { validMove =>
  il match interno che confronta le mosse e dà il risultato
}.getOrElse(DumbUser)
```

in realtà secondo me è più leggibile tenere il match interno qui, dico solo perché è un pattern che vedrai in giro